ledger/src/blockstore.rs

```diff
@@ -2031,6 +2031,33 @@ impl Blockstore {
         slot: Slot,
         require_previous_blockhash: bool,
     ) -> Result<VersionedConfirmedBlock> {
+        self.get_complete_block_with_entries(slot, require_previous_blockhash, false)
+            .map(|result| result.block)
+    }
+
+    pub fn get_rooted_block_with_entries(
+        &self,
+        slot: Slot,
+        require_previous_blockhash: bool,
+    ) -> Result<VersionedConfirmedBlockWithEntries> {
+        datapoint_info!(
+            "blockstore-rpc-api",
+            ("method", "get_rooted_block_with_entries", String)
+        );
+        let _lock = self.check_lowest_cleanup_slot(slot)?;
+
+        if self.is_root(slot) {
+            return self.get_complete_block_with_entries(slot, require_previous_blockhash, true);
+        }
+        Err(BlockstoreError::SlotNotRooted)
+    }
+
+    fn get_complete_block_with_entries(
+        &self,
+        slot: Slot,
+        require_previous_blockhash: bool,
+        populate_entries: bool,
```

Comment:
Makes sense to skip the extra work if we won't use it, and to use a flag to optionally populate / avoid iterating through a second time. Since this field will be optionally populated, should:
```rust
pub entries: Vec<EntrySummary>,
```
be an `Option<Vec<EntrySummary>>` in `VersionedConfirmedBlockWithEntries` ?

This function returns a "complete" block; a complete block must be non-empty which means it will have at least one entry, which also means this vector will never be empty if `populate_entries` is `true`. Based on this reasoning, I could see some argument for leaving it as a regular `Vec<_>` and not an `Option<Vec<_>>`. But, an `Option<_>` is also more expressive of the idea that this Vec was / was not populated.

Reply:
> this vector will never be empty if `populate_entries` is `true`

That isn't how I've written it, though.
I supposed changing it to an Option or renaming the type to `VersionedConfirmedBlockWithMaybeEntries` could make it more obvious that `entries` might not be populated. But maybe I just add a comment somewhere instead?

Reply:
After our DM exchange, I'm onboard to leave off the `Option`.

On my first pass, I failed to notice that this function, `get_complete_block_with_entries()`, is a non-public helper. As such, we don't need to force the `Option<_>` into the struct given that we only return `VersionedConfirmedBlockWithEntries` in the case when `populate_entries == true`. When `populate_entries == false`, we just return the `VersionedConfirmedBlock`.

This is obviously how you designed it but as mentioned, I failed to pick up that this was a helper and not public. So, marking this one resolved with no action